Code/src/test/java/nl/utwente/group10/haskell/expr/ApplyTest.java

```diff
@@ -1,64 +1,95 @@
 package nl.utwente.group10.haskell.expr;
 
-import nl.utwente.group10.haskell.exceptions.HaskellException;
+import nl.utwente.group10.haskell.env.Env;
+import nl.utwente.group10.haskell.exceptions.HaskellTypeError;
+import nl.utwente.group10.haskell.hindley.GenSet;
 import nl.utwente.group10.haskell.type.*;
 import org.junit.Before;
 import org.junit.Test;
 
-import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.*;
 
 public class ApplyTest {
-    private Type T_integer;
-    private Type T_float;
-    private Type T_string;
-    private Type T_numeric;
-    private Type T_a;
-    private Type T_b;
-    private Type T_c;
+    private final Type alpha = new VarT("a");
+    private final Type beta = new VarT("b");
+    private final Type gamma = new VarT("c");
+    private final Type alphaList = new ListT(this.alpha);
+    private final Type betaList = new ListT(this.beta);
+    private final Type integer = new ConstT("Int");
+    private final Type integerList = new ListT(this.integer);
+    private final Type string = new ConstT("Str");
+    private final Type stringList = new ListT(this.string);
+
+    private Env env;
+    private GenSet genSet;
 
     @Before
-    public void setUp() {
-        // Some primitive types
-        this.T_integer = new ConstT("Integer");
-        this.T_float = new ConstT("Float");
-        this.T_string = new ConstT("String");
-
-        // A typeclass
-        this.T_numeric = new TypeClass("Num", T_integer, T_float);
-
-        // Two variable types
-        this.T_a = new VarT("a", T_numeric);
-        this.T_b = new VarT("b", T_numeric);
-        this.T_c = new VarT("c", T_numeric, T_string);
+    public final void setUp() {
+        this.env = new Env();
+        this.genSet = new GenSet();
+
+        this.env.put("id", new FuncT(this.gamma, this.gamma));
+        this.env.put("(+)", new FuncT(this.integer, new FuncT(this.integer, this.integer)));
+        this.env.put("map", new FuncT(new FuncT(this.alpha, this.beta), new FuncT(this.alphaList, this.betaList)));
+        this.env.put("zip", new FuncT(this.alphaList, new FuncT(this.betaList, new ListT(new TupleT(this.alpha, this.beta)))));
+        this.env.put("lcm", new FuncT(this.alpha, new FuncT(this.alpha, this.alpha)));
+    }
+
+    @Test
+    public final void testId() throws HaskellTypeError {
+        final Apply apply = new Apply(new Ident("id"), new Value(this.integer, "42"));
+
+        assertEquals("(id 42)", apply.toHaskell());
+        assertEquals(this.integer.toHaskellType(), apply.analyze(this.env, this.genSet).prune().toHaskellType());
```

Comment:
A helper method `assertHaskellTypeEquals` that does the relevant analyzing, pruning and toHaskellTypeification might be neat.